packages/keyring-controller/src/KeyringController.test.ts

```diff
@@ -3213,6 +3213,39 @@ describe('KeyringController', () => {
       });
     });
   });
+
+  describe('atomic operations', () => {
+    describe('addNewKeyring', () => {
+      it('should rollback the controller keyrings if the keyring creation fails', async () => {
+        const mockAddress = '0x1234';
+        stubKeyringClassWithAccount(MockKeyring, mockAddress);
+        // Mocking the serialize method to throw an error will
+        // halt the controller everytime it tries to persist the keyring,
+        // making it impossible to update the vault
+        jest
+          .spyOn(MockKeyring.prototype, 'serialize')
+          .mockImplementation(async () => {
+            throw new Error('You will never be able to persist me!');
+          });
+        await withController(
+          { keyringBuilders: [keyringBuilderFactory(MockKeyring)] },
+          async ({ controller, initialState }) => {
+            await expect(
+              controller.addNewKeyring(MockKeyring.type),
+            ).rejects.toThrow('You will never be able to persist me!');
+
+            expect(controller.state.keyrings).toHaveLength(1);
```

Comment:
Nit: Perhaps we should repeat this assertion before the `addNewKeyring` step as well, to make clear that there is already a keyring present prior to `addNewKeyring` being called. At first when I read this it seemed like the rollback failed because there is a keyring here.

Reply:
Good point, I changed this to
```typescript
expect(controller.state).toStrictEqual(initialState);
```
So that it's aligned with others, and the intent should be more clear